### PR TITLE
oauth: support authentication methods for client to oauth server requests

### PR DESCRIPTION
**Pull request by Amila-Rukshan**
Commit Message: 
oauth: support authentication methods for client for auth server requests

Additional Description:
Added a new config property under `oauth` filter to define how the `client_id` and `client_secret` will be sent in requests to OAuth server.
```
config:
  ...
  auth_type: "BASIC_AUTH" // defaults to "URL_ENCODED_BODY"
  ...
```
Risk Level:
Testing:
Tests added

Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]

Fixes: #22009 

[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
@snowp WDYT about this? Seems reasonable (even given the dangers of basic auth, if the user is aware of them and the environment is safe). But, there is some nagging thought that we should have some standard way to express HTTP fetch auth etc. for sidecalls. Thoughts?

**Comment by htuch** [comment]
@Amila-Rukshan were you intending for this to be a breaking change, i.e. to start doing basic auth by default? By setting the enum 0 value to this, you create a new default behavior.

@snowp thanks, makes sense. I'm thinking basically anytime we make an HTTP call that a variety of auth methods would be plausibly useful, everything from stats dumps to xDS fetch over HTTP. The said, the fact this is the first time this has come up means this is probably not super useful.

I just noticed https://github.com/envoyproxy/envoy/issues/21851. If we were to implement this we could just leverage it with the OAuth filter making outbound HTTP calls that pick this up.

**Comment by Amila-Rukshan** [comment]
@htuch, yes I wanted to change the existing default method to BASIC_AUTH as it is recommended in the RFC spec. But given that there are people using this filter will have effect on that change, shall I update the PR with existing auth method URL_ENCODED_BODY as the default?

**Comment by rgs1** [comment]
cc: @fishcakez @JuniorHsu since Pinterest might be interested in any breaking behavior here.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto [hunk 1]
```diff
@@ -62,8 +62,17 @@ message OAuth2Credentials {
 
 // OAuth config
 //
-// [#next-free-field: 11]
+// [#next-free-field: 12]
 message OAuth2Config {
+  enum AuthType {
+    // The client_id and client_secret will be sent using HTTP Basic authentication scheme.
+    BASIC_AUTH = 0;
```
Comment (rgs1) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:69]: following up on what @htuch said, we probably want the default value to be a no-op (e.g.: `NO_AUTH` or such).

**Review by rgs1 — CHANGES_REQUESTED** [review]
Sounds like a nice thing to have, but I'd strongly suggest we make the current behavior the default one.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto (hunk 1 above)
Comment (Amila-Rukshan) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto]: @rgs1 @htuch, I have made the existing behaviour as the default auth method (`URL_ENCODED_BODY`).

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto [hunk 2]
```diff
@@ -62,8 +62,17 @@ message OAuth2Credentials {
 
 // OAuth config
 //
-// [#next-free-field: 11]
+// [#next-free-field: 12]
 message OAuth2Config {
+  enum AuthType {
+    // The client_id and client_secret will be sent in the URL encoded request body.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:68]: The other alternative here is making thie DEFAUlT = 0 and allowing that to indicate "client specific default behavior" from the server. That way we could could safely change the Envoy defualt via runtime later, without breaking the wire protoocl. Not sure if there is a strong case for this, so this is just a suggestion.

Comment (htuch) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:68]:
```suggestion 
    // The ``client_id`` and ``client_secret`` will be sent in the URL encoded request body.
```

**Review by htuch — COMMENTED** [review]
/lgtm api

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto (hunk 2 above)
Comment (Amila-Rukshan) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto]: @htuch, are you suggesting in the DEFAULT = 0 case we have another config where we have auth type defined for different auth providers. If that is the case we might need new "auth provider" field in the api do support it. Am I getting you right?

File: source/extensions/filters/http/oauth2/oauth_client.cc [hunk 3]
```diff
@@ -21,21 +22,38 @@ namespace HttpFilters {
 namespace Oauth2 {
 
 namespace {
-constexpr const char* GetAccessTokenBodyFormatString =
+constexpr const char* GetAccessTokenBodyFormatString_01 =
```
Comment (snowp) [inline source/extensions/filters/http/oauth2/oauth_client.cc:25]: Can you give these two a more explanatory name than 1/2?

File: source/extensions/filters/http/oauth2/oauth_client.cc [hunk 4]
```diff
@@ -21,21 +22,38 @@ namespace HttpFilters {
...
+  std::string body;
+  switch (auth_type) {
+  case AuthType::UrlEncodedBody:
+    body = fmt::format(GetAccessTokenBodyFormatString_01, auth_code, encoded_client_id,
+                       encoded_secret, encoded_cb_url);
+    break;
+  case AuthType::BasicAuth:
+    const auto basic_auth_token = absl::StrCat(client_id, ":", secret);
+    const auto encoded_token = Base64::encode(basic_auth_token.data(), basic_auth_token.size());
+    const auto basic_auth_header_value = absl::StrCat("Basic ", encoded_token);
```
Comment (snowp) [inline source/extensions/filters/http/oauth2/oauth_client.cc:50]: Are we able to pre-encode this early on the way we do for the other parameters? That would save us doing this on every request

**Review by snowp — CHANGES_REQUESTED** [review]
At a high level this seems good to me, a few comments

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto]: No, just that we could add DEFAULT = 0 and let the implementatoin pick between 1 and 2.

Comment (Amila-Rukshan) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto]: @htuch, Now I got you. even though auth server supports both 1 and 2 if a request includes auth information in header and encoded body then auth server is supposed to fail (as a validation step). In that way we should avoid sending auth credentials in both types under a default config.

File: source/extensions/filters/http/oauth2/oauth_client.cc [hunk 5]
```diff
@@ -37,20 +37,21 @@ void OAuth2ClientImpl::asyncGetAccessToken(const std::string& auth_code,
   const auto encoded_secret = Http::Utility::PercentEncoding::encode(secret, ":/=&?");
   const auto encoded_cb_url = Http::Utility::PercentEncoding::encode(cb_url, ":/=&?");
 
+  const auto basic_auth_token = absl::StrCat(client_id, ":", secret);
+  const auto encoded_token = Base64::encode(basic_auth_token.data(), basic_auth_token.size());
+  const auto basic_auth_header_value = absl::StrCat("Basic ", encoded_token);
+
```
Comment (snowp) [inline source/extensions/filters/http/oauth2/oauth_client.cc:43]: Ah for some reason I thought we were pre-encoding the data here during filter creation. The way it is now this probably should be inside the switch case. Similarly, I think values that are only used by the UrlEncodedBody should be inside that case block, leaving only shared ones (encoded_cb_url, secret) on top. Sorry for the back and forth here!

Comment (Amila-Rukshan) [reply source/extensions/filters/http/oauth2/oauth_client.cc]: @snowp updated the code. `encoded_cb_url` is common for both the cases

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**